### PR TITLE
TEST-156: int64 => uint64 

### PR DESCRIPTION
**Pull request by riley-stride**
## What is the purpose of the change

1. Replace sf13/cast with our custom fork that checks for overflows
2. Throughout the codebase, use `cast.To__E` from our fork, rather than `cast.To___`. The `E` suffix indicates that we will return `out,err` rather than just `out`, so that we can catch errors! This makes the code more verbose, but safer. 
3. Change a bunch of int64 types to uint64s for extra safety and to reduce the number of casts we must perform

## Brief Changelog

1. Replace cast library in go.mod
2. Replace `cast.ToInt64()` => `cast.ToInt64E` and `cast.ToUint64` => `cast.ToUint64E` and catch any errors
3. Change the following proto struct attributes from uint64 to int64
- depositrecord epochNumber
- UserRedemptionRecord epochNumber
- depositrecord amount
- icaaccount balance
- MsgRedeemStake amt
- epochTracker nextEpochStartTime
- epochTracker duration
- hostZone stakedBal
4. Add a few helpers in utils.go

## Testing and Verifying
unit and integration tests

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? n/a

## Notes
- we should refactor `AddDelegationToValidator` at some point to take in only uint amounts, and make a separate fn for `ReduceDelegationFromValidator`; currently, `AddDelegationToValidator` is used to both add and reduce, which is (a) confusing and (b) requires funky casting! IMO this is an issue for another PR since it's a non-trivial change

### CODE REVIEW

File: x/stakeibc/keeper/msg_server_redeem_stake_test.go [hunk 1]
```diff
@@ -92,7 +93,7 @@ func (suite *KeeperTestSuite) TestRedeemStakeSuccessful() {
 
 	msg := tc.validMsg
 	user := tc.user
-	redeemAmount := sdk.NewInt(msg.Amount)
+	redeemAmount := sdk.NewInt(cast.ToInt64(msg.Amount))
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_redeem_stake_test.go:96]: flagging this

File: x/stakeibc/keeper/host_zone.go [hunk 2]
```diff
@@ -95,22 +95,12 @@ func (k Keeper) GetAllHostZone(ctx sdk.Context) (list []types.HostZone) {
...
-			} else {
-				absAmt := cast.ToUint64(-amt)
-				if absAmt > val.GetDelegationAmt() {
-					k.Logger(ctx).Error(fmt.Sprintf("Delegation amount %d is greater than validator %s delegation amount %d", absAmt, valAddr, val.GetDelegationAmt()))
-					return false
-				}
-				val.DelegationAmt = val.GetDelegationAmt() - absAmt
-				return true
-			}
+			val.DelegationAmt = val.GetDelegationAmt() + cast.ToUint64(amt)
```
Comment (asalzmann) [inline x/stakeibc/keeper/host_zone.go:102]: have we validated that amt is positive? if it's negative, it'll get zero'd 

https://go.dev/play/p/eeAM0EJLiOj

```
func main() {
	var negative int64
	negative = -1000
	fmt.Println("negative", negative)
	zero := cast.ToUint64(negative)
	fmt.Println("zero", zero)
}
```
outputs
```
negative -1000
zero 0

Program exited.
```

File: x/stakeibc/keeper/ibc_handlers.go [hunk 3]
```diff
@@ -177,9 +177,8 @@ func (k Keeper) HandleAcknowledgement(ctx sdk.Context, modulePacket channeltypes
 		}
 	}
 
-	if recordIdToDelete >= 0 {
-		k.RecordsKeeper.RemoveDepositRecord(ctx, cast.ToUint64(recordIdToDelete))
-	}
+	// no need to check id because error case is handled above
+	k.RecordsKeeper.RemoveDepositRecord(ctx, cast.ToUint64(recordIdToDelete))
```
Comment (asalzmann) [inline x/stakeibc/keeper/ibc_handlers.go:181]: flagging

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 4]
```diff
@@ -42,7 +42,7 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
 
 	// construct desired unstaking amount from host zone
 	coinDenom := "st" + hostZone.HostDenom
-	nativeAmount := sdk.NewDec(msg.Amount).Mul(hostZone.RedemptionRate)
+	nativeAmount := sdk.NewDec(cast.ToInt64(msg.Amount)).Mul(hostZone.RedemptionRate)
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_redeem_stake.go:45]: flagging

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 5]
```diff
@@ -58,7 +58,7 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
 	balance := k.bankKeeper.GetBalance(ctx, sender, coinDenom)
 	k.Logger(ctx).Info(fmt.Sprintf("Redemption issuer IBCDenom balance: %v%s", balance.Amount, balance.Denom))
 	k.Logger(ctx).Info(fmt.Sprintf("Redemption requested redemotion amount: %v%s", inCoin.Amount, inCoin.Denom))
-	if balance.Amount.LT(sdk.NewInt(msg.Amount)) {
+	if balance.Amount.LT(sdk.NewInt(cast.ToInt64(msg.Amount))) {
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_redeem_stake.go:61]: flagging

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 6]
```diff
@@ -98,7 +98,7 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
 	hostZoneUnbonding.UserRedemptionRecords = append(hostZoneUnbonding.UserRedemptionRecords, userRedemptionRecord.Id)
 
 	// Escrow user's balance
-	redeemCoin := sdk.NewCoins(sdk.NewCoin(coinDenom, sdk.NewInt(msg.Amount)))
+	redeemCoin := sdk.NewCoins(sdk.NewCoin(coinDenom, sdk.NewInt(cast.ToInt64(msg.Amount))))
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_redeem_stake.go:101]: flagging

**Review by asalzmann — CHANGES_REQUESTED** [review]
Not sure how this got past gosec, but we need to make sure that
1. numbers we cast to uint64s are not negative
2. numbers we cast to int64s are not g.t. math.MaxInt64

we should run this diff by our auditors

**Comment by asalzmann** [comment]
seems like `cast.To...E` doesn't catch overflows

https://go.dev/play/p/sGGbImzlRWh

not sure why. I do think we can fork this library and pretty trivially add it in

File: x/stakeibc/client/cli/tx_redeem_stake.go [hunk 7]
```diff
@@ -19,7 +20,7 @@ func CmdRedeemStake() *cobra.Command {
 		Short: "Broadcast message redeem-stake",
 		Args:  cobra.ExactArgs(3),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			argAmount, err := cast.ToInt64E(args[0])
+			argAmount, err := cast.ToUint64E(args[0])
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/client/cli/tx_redeem_stake.go:23]: Reverted, per offline convo and new To___E fork

File: x/stakeibc/keeper/hooks.go [hunk 8]
```diff
@@ -322,7 +322,7 @@ func (k Keeper) UpdateRedemptionRates(ctx sdk.Context, depositRecords []recordst
 		}
 
 		// calc redemptionRate = (UB+SB+MA)/stSupply
-		redemptionRate := (sdk.NewDec(undelegatedBalance).Add(sdk.NewDec(stakedBalance)).Add(sdk.NewDec(moduleAcctBalance))).Quo(sdk.NewDec(stSupply))
+		redemptionRate := (sdk.NewDec(cast.ToInt64(undelegatedBalance)).Add(sdk.NewDec(cast.ToInt64(stakedBalance))).Add(sdk.NewDec(cast.ToInt64(moduleAcctBalance)))).Quo(sdk.NewDec(stSupply))
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/hooks.go:325]: Addressed per new cast fork and using cast.To____E

File: x/stakeibc/keeper/host_zone.go (hunk 2 above)
Comment (riley-stride) [reply x/stakeibc/keeper/host_zone.go:102]: Addressed per new cast fork and using cast.To____E

File: x/stakeibc/keeper/ibc_handlers.go (hunk 3 above)
Comment (riley-stride) [reply x/stakeibc/keeper/ibc_handlers.go:181]: Addressed per new cast fork and using cast.To____E

File: x/stakeibc/keeper/msg_server_redeem_stake.go (hunk 4 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:45]: Addressed per new cast fork and using cast.To____E

File: x/stakeibc/keeper/msg_server_redeem_stake.go (hunk 5 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:61]: Addressed per new cast fork and using cast.To____E

File: x/stakeibc/keeper/msg_server_redeem_stake.go (hunk 6 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:101]: Addressed per new cast fork and using cast.To____E

File: x/stakeibc/keeper/msg_server_redeem_stake_test.go (hunk 1 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_redeem_stake_test.go:96]: Addressed per new cast fork and using cast.To____E

**Comment by riley-stride** [comment]
Re-did this PR from scratch [here](https://github.com/Stride-Labs/stride/pull/183). Copying over the description and closing this PR.

**Pull request closed without merging**